File: aerox.Runtime.Widgets/Container.cs
using aerox.Runtime.Extensions;
using aerox.Runtime.Math;
using aerox.Runtime.Widgets.Events;
using aerox.Runtime.Widgets.Graphics;
using aerox.Runtime.Widgets.Graphics.Commands;

namespace aerox.Runtime.Widgets;
public abstract class Container : Widget
{
    private readonly List<Slot> _slot = new();
    
    protected readonly Mutex SlotsMutex = new();
    
    public ClipMode Clip = ClipMode.None;
    
    public Container(IEnumerable<Widget> children)
    {
        foreach (var widget in children)
            if (AddChild(widget) == null)
                break;
    }
    
    public Container(IEnumerable<Slot> children)
    {
        foreach (var slot in children)
            if (AddChild(slot) == null)
                break;
    }
    
    public Container()
    {
    }
    
    protected abstract void ArrangeSlots(Size2d drawSize);

    public virtual void OnChildResized(Widget widget)
    {
        if (CheckSize()) ArrangeSlots(GetContentSize());
    }
    
    public virtual Slot MakeSlot(Widget widget)
    {
        return new Slot(widget,this);
    }
    
    public virtual Slot? AddChild<TE>() where TE : Widget
    {
        return AddChild(Activator.CreateInstance<TE>());
    }
    
    public virtual void OnSlotUpdated(Slot slot)
    {
    }

    public Slot? AddChild(Widget widget)
    {
        lock (SlotsMutex)
        {
            var maxSlots = GetMaxSlots();
            if (maxSlots > 0 && _slot.Count == maxSlots) return null;

            widget.SetParent(this);

            var slot = MakeSlot(widget);
            
            slot.SetOwner(this);
            
            _slot.Add(slot);

            if (Surface != null) widget.NotifyAddedToSurface(Surface);

            OnChildResized(widget);
            Console.WriteLine("Added child [{0}] to container [{1}]", widget.GetType().Name, GetType().Name);
            return slot;
        }
    }
    
    public Slot? AddChild(Slot slot)
    {
        lock (SlotsMutex)
        {
            var maxSlots = GetMaxSlots();
            if (maxSlots > 0 && _slot.Count == maxSlots) return null;
                
            var widget = slot.GetWidget();
            
            widget.SetParent(this);
            
            slot.SetOwner(this);
            
            _slot.Add(slot);

            if (Surface != null) widget.NotifyAddedToSurface(Surface);

            OnChildResized(widget);
            Console.WriteLine("Added child [{0}] to container [{1}]", widget.GetType().Name, GetType().Name);
            return slot;
        }
    }

    public virtual bool RemoveChild(Widget widget)
    {
        lock (SlotsMutex)
        {
            for (var i = 0; i < _slot.Count; i++)
            {
                if (_slot[i].GetWidget() != widget) continue;

                widget.SetParent(null);

                if (Surface != null) widget.NotifyRemovedFromSurface(Surface);

                _slot.RemoveAt(i);

                CheckSize();
                return true;
            }

            return false;
        }
    }

    public virtual Slot? GetSlot(int idx)
    {
        lock (SlotsMutex)
        {
            return idx < _slot.Count ? _slot[idx] : null;
        }
    }

    public virtual Slot[] GetSlots()
    {
        lock (SlotsMutex)
        {
            return _slot.ToArray();
        }
    }

    public virtual int GetNumSlots()
    {
        lock (SlotsMutex)
        {
            return _slot.Count;
        }
    }

    public virtual uint GetMaxSlots() => 0;

    public override Widget? ReceiveCursorDown(CursorDownEvent e, TransformInfo info)
    {
        if (IsChildrenHitTestable())
        {
            var res = ChildrenReceiveCursorDown(e, info);
            if (res != null ) return res;
        }

        return base.ReceiveCursorDown(e, info);
    }

    protected virtual Widget? ChildrenReceiveCursorDown(CursorDownEvent e, TransformInfo info)
    {
        var point = e.Position.Cast<float>();
        lock (SlotsMutex)
        {
            foreach (var slot in _slot.AsReversed())
            {
                var slotInfo = ComputeChildTransform(slot, info);
                if (!slotInfo.PointWithin(point)) continue;
                var res = slot.GetWidget().ReceiveCursorDown(e, slotInfo);
                if (res != null) return res;
            }
        }

        return null;
    }

    public override void ReceiveCursorEnter(CursorMoveEvent e, TransformInfo info, List<Widget> items)
    {
        if (IsChildrenHitTestable()) ChildrenReceiveCursorEnter(e, info, items);

        base.ReceiveCursorEnter(e, info, items);
    }

    protected virtual void ChildrenReceiveCursorEnter(CursorMoveEvent e, TransformInfo info, List<Widget> items)
    {
        var point = e.Position.Cast<float>();
        lock (SlotsMutex)
        {
            foreach (var slot in _slot)
            {
                var slotInfo = ComputeChildTransform(slot, info);
                if (slotInfo.PointWithin(point))
                    slot.GetWidget().ReceiveCursorEnter(e, slotInfo, items);
            }
        }
    }

    public override bool ReceiveCursorMove(CursorMoveEvent e, TransformInfo info)
    {
        if (IsChildrenHitTestable() && ChildrenReceiveCursorMove(e, info)) return true;

        return base.ReceiveCursorMove(e, info);
    }

    protected virtual bool ChildrenReceiveCursorMove(CursorMoveEvent e, TransformInfo info)
    {
        var point = e.Position.Cast<float>();
        lock (SlotsMutex)
        {
            foreach (var slot in _slot)
            {
                var slotInfo = ComputeChildTransform(slot, info);
                if (slotInfo.PointWithin(point) &&
                    slot.GetWidget().ReceiveCursorMove(e, slotInfo))
                    return true;
            }
        }

        return false;
    }

    public override bool ReceiveScroll(ScrollEvent e, TransformInfo info)
    {
        if (IsChildrenHitTestable() && ChildrenReceiveScroll(e, info)) return true;

        return base.ReceiveScroll(e, info);
    }

    protected virtual bool ChildrenReceiveScroll(ScrollEvent e, TransformInfo info)
    {
        var point = e.Position.Cast<float>();
        lock (SlotsMutex)
        {
            foreach (var slot in _slot)
            {
                var slotInfo = ComputeChildTransform(slot, info);
                if (slotInfo.PointWithin(point) &&
                    slot.GetWidget().ReceiveScroll(e, slotInfo))
                    return true;
            }
        }

        return false;
    }

    public override void SetSize(Size2d size)
    {
        base.SetSize(size);
        ArrangeSlots(GetContentSize());
    }

    public override void NotifyAddedToSurface(Surface surface)
    {
        base.NotifyAddedToSurface(surface);
        lock (SlotsMutex)
        {
            foreach (var slot in _slot)
                slot.GetWidget().NotifyAddedToSurface(surface);
        }
    }

    public override void NotifyRemovedFromSurface(Surface surface)
    {
        base.NotifyRemovedFromSurface(surface);
        lock (SlotsMutex)
        {
            foreach (var slot in _slot)
                slot.GetWidget().NotifyRemovedFromSurface(surface);
        }
    }

    protected override void OnDispose(bool isManual)
    {
        base.OnDispose(isManual);
        lock (SlotsMutex)
        {
            foreach (var slot in _slot) slot.GetWidget().Dispose();
        }
    }

    public TransformInfo ComputeChildTransform(Slot slot, TransformInfo info)
    {
        return ComputeChildTransform(slot.GetWidget(), info);
    }

    public virtual TransformInfo ComputeChildTransform(Widget widget, TransformInfo info)
    {
        return new TransformInfo(info.Transform.Translate(new Vector2<float>(Padding.Left,Padding.Top)) * widget.ComputeRelativeTransform(),widget.GetSize(),info.Depth + 1);
    }

    public override void CollectContent(TransformInfo info, DrawCommands drawCommands)
    {
        switch (Clip)
        {
            case ClipMode.None:
                foreach (var slot in GetSlots())
                {
                    var newTransform = ComputeChildTransform(slot, info);
                    var widget = slot.GetWidget();
                    widget.Collect(newTransform,drawCommands);
                }
                break;
            case ClipMode.Bounds:
            {
                drawCommands.PushClip(info,this);

                var myAAR = info.ToRect();
                
                foreach (var slot in GetSlots())
                {
                    var newTransform = ComputeChildTransform(slot, info);
                    var slotAAR = newTransform.ToRect();
                    
                    if(!myAAR.IntersectsWith(slotAAR)) continue;
                    
                    var widget = slot.GetWidget();
                    widget.Collect(newTransform,drawCommands);
                }
                
                drawCommands.PopClip();
            }
                break;
        }
    }
}

File: modules/widgets/lib/aerox/widgets/Container.cpp
#include "aerox/widgets/Container.hpp"

#include "aerox/widgets/ContainerSlot.hpp"
#include "aerox/widgets/event/CursorDownEvent.hpp"
#include "aerox/widgets/event/CursorMoveEvent.hpp"
#include "aerox/widgets/event/ScrollEvent.hpp"

namespace aerox::widgets
{
    void Container::OnChildResized(Widget * widget)
    {
        if(CheckSize()) ArrangeSlots(GetContentSize());
    }

    size_t Container::GetMaxSlots() const
    {
        return std::numeric_limits<size_t>::max();
    }

    size_t Container::GetUsedSlots() const
    {
        return _slots.size();
    }

    Shared<ContainerSlot> Container::GetSlot(int index) const
    {
        return index < _slots.size() ? _slots[index] : Shared<ContainerSlot>{};
    }

    Shared<ContainerSlot> Container::GetSlot(Widget* widget) const
    {
        return _widgetsToSlots.contains(widget) ? _widgetsToSlots.at(widget) : Shared<ContainerSlot>{};
    }

    std::vector<Shared<ContainerSlot>> Container::GetSlots() const
    {
        return _slots;
    }

    Shared<ContainerSlot> Container::AddChild(const Shared<Widget>& widget)
    {
        if(!widget) return {};
        
        auto newSlotCount = _slots.size() + 1;
        
        if(newSlotCount > GetMaxSlots()) return {};

        auto slot = MakeSlot(widget);
        _slots.push_back(slot);
        _widgetsToSlots.emplace(widget.get(),slot);
        widget->SetParent(this->GetSharedDynamic<Container>());
        
        if(auto surf = GetSurface())
        {
            widget->NotifyAddedToSurface(surf);
        }

        OnChildResized(widget.get());
        
        return slot;
    }

    bool Container::RemoveChild(const Shared<Widget>& widget)
    {
        if(!_widgetsToSlots.contains(widget.get())) return false;
        
        for(auto i = 0; i < _slots.size(); i++)
        {
            if(_slots[i]->GetWidget() != widget) continue;
            
            widget->SetParent({});
            
            if(auto surf = GetSurface())
            {
                widget->NotifyRemovedFromSurface(surf);
            }
            
            _slots.erase(_slots.begin() + i);
            _widgetsToSlots.erase(widget.get());

            CheckSize();
            return true;
        }

        return false;
    }

    void Container::SetDrawSize(const Vec2<float>& size)
    {
        Widget::SetDrawSize(size);
        ArrangeSlots(GetContentSize());
    }

    void Container::OnDispose(bool manual)
    {
        Widget::OnDispose(manual);
        
        for (auto &slot : GetSlots())
        {
            slot->GetWidget()->Dispose();
        }

        _slots.clear();
    }

    void Container::NotifyAddedToSurface(const Shared<Surface>& widgetSurface)
    {
        Widget::NotifyAddedToSurface(widgetSurface);
        for(auto &slot : GetSlots())
        {
            slot->GetWidget()->NotifyAddedToSurface(widgetSurface);
        }
    }

    void Container::NotifyRemovedFromSurface(const Shared<Surface>& widgetSurface)
    {
        Widget::NotifyRemovedFromSurface(widgetSurface);
        for(auto &slot : GetSlots())
        {
            slot->GetWidget()->NotifyRemovedFromSurface(widgetSurface);
        }
    }

    Shared<Widget> Container::NotifyCursorDown(const Shared<CursorDownEvent>& event, const TransformInfo& transform)
    {
        if(AreChildrenHitTestable())
        {
            if(auto widget = NotifyChildrenCursorDown(event,transform)) return widget;
        }
        return Widget::NotifyCursorDown(event, transform);
    }

    void Container::NotifyCursorEnter(const Shared<CursorMoveEvent>& event, const TransformInfo& transform,
                                      std::vector<Shared<Widget>>& items)
    {
        if(AreChildrenHitTestable()) NotifyChildrenCursorEnter(event,transform,items);
        
        Widget::NotifyCursorEnter(event, transform, items);
    }

    bool Container::NotifyCursorMove(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
    {
        if(AreChildrenHitTestable()) return NotifyChildrenCursorMove(event,transform);
        return Widget::NotifyCursorMove(event, transform);
    }

    bool Container::NotifyScroll(const Shared<ScrollEvent>& event, const TransformInfo& transform)
    {
        if(AreChildrenHitTestable()) return NotifyChildrenScroll(event,transform);
        return Widget::NotifyScroll(event, transform);
    }

    Shared<Widget> Container::NotifyChildrenCursorDown(const Shared<CursorDownEvent>& event,
        const TransformInfo& transform)
    {
        for (auto &slot : GetSlots())
        {
            auto slotTransform = ComputeChildTransform(slot,transform);
            if(slotTransform.IsPointWithin(event->position) && slot->GetWidget()->NotifyCursorDown(event,slotTransform)) return slot->GetWidget();
        }
        return {};
    }

    bool Container::NotifyChildrenCursorEnter(const Shared<CursorMoveEvent>& event, const TransformInfo& transform, std::vector<Shared<Widget>>& items)
    {
        for (auto &slot : GetSlots())
        {
            auto slotTransform = ComputeChildTransform(slot,transform);
            if(slotTransform.IsPointWithin(event->position))
            {
                slot->GetWidget()->NotifyCursorEnter(event,slotTransform,items);
            }
        }
        return false;
    }

    bool Container::NotifyChildrenCursorMove(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
    {
        for (auto &slot : GetSlots())
        {
            auto slotTransform = ComputeChildTransform(slot,transform);
            if(slotTransform.IsPointWithin(event->position) && slot->GetWidget()->NotifyCursorMove(event,slotTransform)) return true;
        }
        return false;
    }

    bool Container::NotifyChildrenScroll(const Shared<ScrollEvent>& event, const TransformInfo& transform)
    {
        for (auto &slot : GetSlots())
        {
            auto slotTransform = ComputeChildTransform(slot,transform);
            if(slotTransform.IsPointWithin(event->position) && slot->GetWidget()->NotifyScroll(event,slotTransform)) return true;
        }
        return false;
    }

    void Container::Collect(const TransformInfo& transform, std::vector<Shared<DrawCommand>>& drawCommands)
    {
        for (auto &slot : GetSlots())
        {
            TransformInfo t = ComputeChildTransform(slot,transform);
            auto widget = slot->GetWidget();
            widget->Collect(t,drawCommands);
        } 
    }

    TransformInfo Container::ComputeChildTransform(const Shared<ContainerSlot>& slot, const TransformInfo& myTransform)
    {
        return ComputeChildTransform(slot->GetWidget(),myTransform);
    }

    TransformInfo Container::ComputeChildTransform(const Shared<Widget>& widget, const TransformInfo& myTransform)
    {
        return TransformInfo{myTransform.transform.Translate(Vec2(padding.left,padding.top)) * widget->ComputeRelativeTransform(),widget->GetDrawSize(),myTransform.depth + 1};
    }
}
